Additional-work-№2-from-11.04.2023/main.c

```diff
@@ -0,0 +1,231 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stddef.h>
+
+int *readFromFile(char *nameOfTheFile, int *length)
+{
+    FILE *file = fopen(nameOfTheFile, "r");
+    int amount = 0;
+    fscanf(file, "%d", &amount);
+    int *array = calloc(amount, sizeof(int));
+    for (int i = 0; i < amount; ++i)
+    {
+        fscanf(file, "%d", &array[i]);
+    }
+    fclose(file);
+    *length = amount;
+    return array;
+}
+
+/*
+ * for debug
+void printArray(int *array, int arrayLength)
+{
+    for (int i = 0; i < arrayLength; ++i)
+    {
+        printf("%d ", array[i]);
+    }
+    printf("\n");
+}
+*/
+
+void swap(int *firstElement, int *secondElement)
+{
+    int temp = *secondElement;
+    *secondElement = *firstElement;
+    *firstElement = temp;
+}
+
+int divide(int *array, int left, int right)
+{
+    int changingElem = left;
+    for (int i = left; i < right; ++i)
+    {
+        if (array[i] < array[right])
+        {
+            swap(&array[changingElem], &array[i]);
+            changingElem += 1;
+        }
+    }
+    swap(&array[right], &array[changingElem]);
+    return changingElem;
+}
+
+void quickSortRec(int *array, int left, int right)
+{
+    if (left < right)
+    {
+        int centerElem = divide(array, left, right);
+        quickSortRec(array, left, centerElem - 1);
+        quickSortRec(array, centerElem + 1, right);
+    }
+}
+
+void quickSort(int *array, int lenOfArray)
+{
+    quickSortRec(array, 0, lenOfArray - 1);
+}
+
+int *intersection(int *firstArray, int lenOfTheFirstArray, int *secondArray, int lenOfTheSecondArray, int *lenOfIntersectionArray)
+{
+    int firstPointer = 0, secondPointer = 0, interArrayPointer = 0;
+    int maxLen = lenOfTheFirstArray * (lenOfTheFirstArray >= lenOfTheSecondArray) + lenOfTheSecondArray * (lenOfTheSecondArray > lenOfTheFirstArray);
+    int *intersectionArray = calloc(maxLen, sizeof(int));
+    if (intersectionArray == NULL)
+    {
+        return NULL;
+    }
+    while (firstPointer != lenOfTheFirstArray && secondPointer != lenOfTheSecondArray)
+    {
+        if (firstArray[firstPointer] == secondArray[secondPointer])
+        {
+            intersectionArray[interArrayPointer] = firstArray[firstPointer];
+            firstPointer++;
+            interArrayPointer++;
+        }
+        else if (firstArray[firstPointer] < secondArray[secondPointer])
+        {
+            firstPointer++;
+        }
+        else
+        {
+            secondPointer++;
+        }
+    }
+    *lenOfIntersectionArray = interArrayPointer;
+    return intersectionArray;
+}
+
+int *unification(int *firstArray, int lenOfTheFirstArray, int *secondArray, int lenOfTheSecondArray, int *lenOfUnificationArray)
```

Comment:
```suggestion
int *union(int *firstArray, int lenOfTheFirstArray, int *secondArray, int lenOfTheSecondArray, int *lenOfUnificationArray)
```
Так принято объединение называть ([proof](https://en.wikipedia.org/wiki/Union_(set_theory))).

Reply:
Понятно. `union` - ключевое слово в си. Можно `arraysUnion`.